networkx/algorithms/bipartite/extendability.py

```diff
@@ -0,0 +1,133 @@
+    graph $G_M$ is strongly connected and there are $k$ vertex-disjoint directed
+    paths between every vertex of U and every vertex of V.
+
+    Assuming that input graph G is undirected, simple, connected, bipartite and contains
+    a perfect matching M, this function constructs the residual graph $G_M$ of G and
+    returns the minimum value among the maximum vertex-disjoint directed paths between
+    every vertex of U and every vertex of V in $G_M$. By combining the second definition
+    and the lemma, notice that this value represents the extendability of the graph G.
+
+    Parameters
+    ----------
+
+        G : NetworkX Graph
+
+    Returns
+    -------
+
+        extendability: int
+
+    Raises
+    ------
+
+        NetworkXError
+           If the graph G is not simple.
+           If the graph G is disconnected.
+           If the graph G is not bipartite.
+           If the graph G does not contain a perfect matching.
+           If the residual graph of G is not strongly connected.
+
+    Notes
+    -----
+
+        Time complexity O($n^3$ $m^2$)) where $n$ is the number of vertices
+        and $m$ is the number of edges.
+
+    References
+    ----------
+
+        ..[1] "A polynomial algorithm for the extendability problem in bipartite graphs",
+              J. Lakhal, L. Litzler, Information Processing Letters, 1998.
+
+    """
+
+    for edge in G.edges:
+        if edge[0] == edge[1]:
+            raise nx.NetworkXError("Graph G is not simple")
+
+    if not nx.is_connected(G):
+        raise nx.NetworkXError("Graph G is not connected")
+
+    if not nx.bipartite.is_bipartite(G):
+        raise nx.NetworkXError("Graph G is not bipartite")
+
+    U, V = nx.bipartite.sets(G)
+
+    # Variable $k$ stands for the extendability of graph G
+    k = float("Inf")
+
+    maximum_matching = nx.bipartite.hopcroft_karp_matching(G)
+
+    if not nx.is_perfect_matching(G, maximum_matching):
+        raise nx.NetworkXError("Graph G does not contain a perfect matching")
+
+    # Construct a list consists of the edges of M by directing them from V to U
+    perfect_matching = []
+    for vertex in maximum_matching.keys():
+        if vertex in V:
+            neighbor = maximum_matching[vertex]
+            perfect_matching.append((vertex, neighbor))
+
+    # Direct all the edges of G
+    directed_edges = []
+    for edge in G.edges:
+        first_coordinate, second_coordinate = edge[0], edge[1]
+        reverse_pair = (second_coordinate, first_coordinate)
+        if first_coordinate in U and second_coordinate in V:
+            if reverse_pair in perfect_matching:
+                directed_edges.append(reverse_pair)
+            else:
+                directed_edges.append(edge)
+        if first_coordinate in V and second_coordinate in U:
+            if edge in perfect_matching:
+                directed_edges.append(edge)
+            else:
+                directed_edges.append(reverse_pair)
```

Comment:
The logic here is pretty complicated - can it be simplified? The comment says "direct all edges of G" - if that's really what's being done, why not use `G.to_directed().edges`? If you only want the directed edges that are in the perfect matching, how about:

```python
directed_edges = [edge for edge in G.to_directed().edges if edge in perfect_matching]
```